// hy-component/es5/rxjs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createXObservable = exports.rxjsMixin = void 0;

var _esm = require("rxjs/_esm5");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var rxjsMixin = function rxjsMixin(C) {
  return (
    /*#__PURE__*/
    function (_C) {
      _inherits(_class, _C);

      function _class() {
        _classCallCheck(this, _class);

        return _possibleConstructorReturn(this, _getPrototypeOf(_class).apply(this, arguments));
      }

      _createClass(_class, [{
        key: "setupComponent",
        value: function setupComponent(el, opts) {
          var _this = this;

          var sideEffects = {};
          this.subjects = {};
          this.subjects.disconnect = new _esm.Subject();
          this.subjects.document = new _esm.ReplaySubject();
          Object.keys(this.constructor.types).map(function (key) {
            _this.subjects[key] = new _esm.ReplaySubject(1);

            sideEffects[key] = function (x) {
              return _this.subjects[key].next(x);
            };
          });
          Object.defineProperty(this.constructor, "sideEffects", {
            get: function get() {
              return sideEffects;
            },
            set: function set() {},
            enumerable: true,
            configurable: true
          });

          _get(_getPrototypeOf(_class.prototype), "setupComponent", this).call(this, el, opts);
        }
      }, {
        key: "connectComponent",
        value: function connectComponent() {
          var _this2 = this;

          _get(_getPrototypeOf(_class.prototype), "connectComponent", this).call(this);

          this.subjects.document.next(document);
          Object.keys(this.constructor.types).map(function (key) {
            return _this2.subjects[key].next(_this2[key]);
          });
        }
      }, {
        key: "disconnectComponent",
        value: function disconnectComponent() {
          _get(_getPrototypeOf(_class.prototype), "disconnectComponent", this).call(this);

          this.subjects.disconnect.next({});
        }
      }, {
        key: "adaptComponent",
        value: function adaptComponent() {
          _get(_getPrototypeOf(_class.prototype), "adaptComponent", this).call(this);

          this.subjects.document.next(document);
        }
      }]);

      return _class;
    }(C)
  );
};

exports.rxjsMixin = rxjsMixin;

var createXObservable = function createXObservable(X) {
  return function (el, cOpts, oOpts) {
    return _esm.Observable.create(function (obs) {
      var next = obs.next.bind(obs);
      var observer = new X(function (xs) {
        return Array.from(xs).forEach(next);
      }, cOpts);
      /* if (process.env.DEBUG) console.log("observe", X.name); */

      observer.observe(el, oOpts);
      return function () {
        /* if (process.env.DEBUG) console.log("unobserve", X.name); */
        observer.unobserve(el);
      };
    });
  };
};

exports.createXObservable = createXObservable;
