// hy-component/es5/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "array", {
  enumerable: true,
  get: function get() {
    return _attrTypes.array;
  }
});
Object.defineProperty(exports, "arrayOf", {
  enumerable: true,
  get: function get() {
    return _attrTypes.arrayOf;
  }
});
Object.defineProperty(exports, "bool", {
  enumerable: true,
  get: function get() {
    return _attrTypes.bool;
  }
});
Object.defineProperty(exports, "number", {
  enumerable: true,
  get: function get() {
    return _attrTypes.number;
  }
});
Object.defineProperty(exports, "oneOf", {
  enumerable: true,
  get: function get() {
    return _attrTypes.oneOf;
  }
});
Object.defineProperty(exports, "regex", {
  enumerable: true,
  get: function get() {
    return _attrTypes.regex;
  }
});
Object.defineProperty(exports, "string", {
  enumerable: true,
  get: function get() {
    return _attrTypes.string;
  }
});
exports.default = void 0;

var _attrTypes = require("attr-types");

var _default = {
  array: _attrTypes.array,
  arrayOf: _attrTypes.arrayOf,
  bool: _attrTypes.bool,
  number: _attrTypes.number,
  oneOf: _attrTypes.oneOf,
  regex: _attrTypes.regex,
  string: _attrTypes.string
};
exports.default = _default;
